File: 1114-print-in-order/1114-print-in-order.cpp
class Foo {
public:
    mutex a,b;
    Foo() {
        a.lock(),b.lock();
    }

    void first(function<void()> printFirst) {
        
        // printFirst() outputs "first". Do not change or remove this line.
        printFirst();
        a.unlock();
    }

    void second(function<void()> printSecond) {
        lock_guard<mutex> g(a);
        // printSecond() outputs "second". Do not change or remove this line.
        printSecond();
        b.unlock();
    }

    void third(function<void()> printThird) {
        // b.lock();
        lock_guard<mutex>g(b);
        // printThird() outputs "third". Do not change or remove this line.
        printThird();
        b.unlock();
    }
};

File: 1114-print-in-order/1114-print-in-order.py
from threading import Lock
class Foo:
    def __init__(self):
        self.lock1,self.lock2 = Lock(),Lock()
        self.lock1.acquire()
        self.lock2.acquire()


    def first(self, printFirst: 'Callable[[], None]') -> None:
        
        # printFirst() outputs "first". Do not change or remove this line.
        printFirst()
        self.lock1.release()


    def second(self, printSecond: 'Callable[[], None]') -> None:
        self.lock1.acquire()
        # printSecond() outputs "second". Do not change or remove this line.
        printSecond()
        self.lock2.release()


    def third(self, printThird: 'Callable[[], None]') -> None:
        self.lock2.acquire()
        # printThird() outputs "third". Do not change or remove this line.
        printThird()